hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -29,14 +40,14 @@ def transform(self, model, node):
             if type_name not in inferred_types:
                 self._infer_default_type(node, type_name)
 
-        return False  # No model graph changes made
+        return True  # May need to rerun
```

Comment:
Why? This optimizer doesn't add or remove nodes. Others that do and have this optimizer in the flow will have it rerun. If you have an optimizer that inserts new nodes or replaces nodes with ones that need type inference and has no way of stopping the problem is there, not here. Usually we solve this with adding a hidden attribute that the optimizer checks so it doesn't run into an infinite loop. Check for example how `resource_strategy.py` optimizer prevents itself from running over and over again

Reply:
One cannot set the output type if the input type is unknown. Thus, the inferred precisions have to propagate through the system as output types are set, which subsequent become input types for the next layer. There is an order one has to follow. However, there is no guarantee that we iterate through layers in the needed order, particularly if there are branches, so this is to make sure we try running the optimizer on all layers again. We don't have to run other optimizes from the beginning, just this one, but I am not sure if that can be specified.